include/mbedtls/ssl.h

```diff
@@ -2291,6 +2355,85 @@ int mbedtls_ssl_conf_alpn_protocols( mbedtls_ssl_config *conf, const char **prot
+ *                          (Default: MBEDTLS_SSL_DTLS_SRTP_MKI_UNSUPPORTED)
+ *
+ * \param conf              SSL configuration
+ * \param support_mki_value Enable or disable (MBEDTLS_SSL_DTLS_SRTP_MKI_UNSUPPORTED
+ *                          or MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED)
+ */
+void mbedtls_ssl_conf_srtp_mki_value_supported( mbedtls_ssl_config *conf,
+                                                int support_mki_value );
+
+/**
+ * \brief                   Set the supported DTLS-SRTP protection profiles.
+ *
+ * \param conf              SSL configuration
+ * \param profiles          List of supported protection profiles,
+ *                          in decreasing preference order.
+ * \param profiles_number   Number of supported profiles.
+ *
+ * \return         0 on success, or MBEDTLS_ERR_SSL_BAD_INPUT_DATA.
+ */
+int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
+                                                    const mbedtls_ssl_srtp_profile *profiles,
+                                                    size_t profiles_number );
+
+/**
+ * \brief                   Set the mki_value for the current dtls session.
+ *
+ * \param ssl              SSL context
+ * \param mki_value        MKI value to set
+ * \param mki_len          MKI length
+ *
+ * \return         0 on success, MBEDTLS_ERR_SSL_BAD_INPUT_DATA
+ *                 or MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE on failure
+ */
+int mbedtls_ssl_dtls_srtp_set_mki_value( mbedtls_ssl_context *ssl,
+                                         unsigned char *mki_value,
+                                         size_t mki_len );
+/**
+ * \brief          Get the negotiated DTLS-SRTP Protection Profile.
+ *                 This function should be called after the handshake is
+ *                 completed.
+ *
+ * \param ssl      SSL context
+ *
+ * \return         Protection Profile enum member,
+ *                 MBEDTLS_SRTP_UNSET_PROFILE if no protocol was negotiated.
+ */
+mbedtls_ssl_srtp_profile mbedtls_ssl_get_dtls_srtp_protection_profile( const mbedtls_ssl_context *ssl );
+
+/**
+ * \brief                  Get the generated DTLS-SRTP key material.
+ *                         This function should be called after the handshake is
+ *                         completed. It shall returns 80 bytes of key material
+ *                         generated according to RFC5764
+ *
+ * \param ssl              SSL context
+ * \param key              Buffer to hold the generated key material
+ * \param key_len          [in/out] key buffer size. outputs the actual number
+ *                         of bytes written
+ *
+ * \return         0 on succes, MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL if the key buffer
+ *                 is too small to hold the generated key
+ */
+int mbedtls_ssl_get_dtls_srtp_key_material( const mbedtls_ssl_context *ssl,
+                                            unsigned char *key,
+                                            size_t *key_len );
```

Comment:
ok, but I really don't like this style

Reply:
@RonEld With 3.0 coming up, we have the chance to change this throughout the library, and I think it would be worth brining up this point in an architecture discussion.

Reply:
I think it's definitely worth considering, e.g. because it reduces the number of arguments and therefore makes it more likely that compilers can pass them in registers as opposed to stack.

Reply:
Ping @mpg

Reply:
I think in/out parameters are bad for readability (at least for lengths), so I've actively tried to remove them from our API when moving from 1.3.x to 2.0, as I think readability is extremely important when it comes to buffer sizes. (In practice, their was a case where I initially misread some of our code, then during review Paul misread it in the exact same way, and my conclusion was that it was the in/out length that tripped both of us, hence they were bad.) I've even gone as far as embed this point of view in our [coding standard][1] (search for "limited use of in-out parameters" as we can't link to a section).

[1]: https://tls.mbed.org/kb/development/mbedtls-coding-standards

I'm happy to discuss that again in an architecture discussion for 3.0, but for now I'd rather stick to the current coding standards - as Hanno says, uniformity with the rest of our API matters, if we change it let's do it all at once and after we had a discussion.

(FWIW, some people from the crypto team seem to agree with me, as they are systematically using `uint8_t *, size_t, size_t *` in the signature of functions that write to buffers.)

Reply:
AFAIK, this API in crypto library is used because that's what's used in the tls library, and I have the same argument with the person that agrees with you.

FWIW, when I see a function with a pointer to size as a parameter, I *always* expect it to be initialized with the buffer size. In addition, I am always in favor of reducing number of parameters, as this is much more readable, IMHO, and as @hanno-arm mentioned, compilers can pass them in registers, as opposed to stack.
Of course, for uniformity, I will change the API, but I really think we should consider this for version 3.0

Reply:
Thanks for agreeing to change for uniformity for now, and I agree this is a discussion we need to have, as on the readability front we don't have agreement within the team, and even for people who think the 3-argument version is more readable (as I do), there is the point that it might impact stack usage, speed and/or code size, which may or may not outweigh the perceived readability gains.